go/tools/gazelle/rules/generator.go

```diff
@@ -148,26 +158,81 @@ func (g *generator) generateBin(rel, library string, pkg *build.Package) (*bzl.R
 
 }
 
-func (g *generator) generateLib(rel, name string, pkg *build.Package) (*bzl.Rule, error) {
+func (g *generator) generateLib(rel, name string, pkg *build.Package, cgoName string) (*bzl.Rule, error) {
 	kind := "go_library"
 
 	visibility := "//visibility:public"
 	// Libraries made for a go_binary should not be exposed to the public.
 	if pkg.IsCommand() {
 		visibility = "//visibility:private"
 	}
+
+	attrs := []keyvalue{
+		{key: "name", value: name},
+	}
+
+	if len(cgoName) == 0 {
+		srcs := append([]string{}, pkg.GoFiles...)
+		srcs = append(srcs, pkg.SFiles...)
+		attrs = append(attrs, keyvalue{key: "srcs", value: srcs})
+	} else {
+		// go_library gets mad when an empty slice is passed in, but handles not
+		// being set at all just fine when "library" is set.
+		if len(pkg.GoFiles) != 0 {
+			attrs = append(attrs, keyvalue{key: "srcs", value: pkg.GoFiles})
+		}
+		attrs = append(attrs, keyvalue{key: "library", value: ":" + cgoName})
+	}
 	visibility = checkInternalVisibility(rel, visibility)
 
+	attrs = append(attrs, keyvalue{key: "visibility", value: []string{visibility}})
+
+	deps, err := g.dependencies(pkg.Imports, rel)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(deps) > 0 {
+		attrs = append(attrs, keyvalue{key: "deps", value: deps})
+	}
+
+	return newRule(kind, nil, attrs)
+}
+
+// Generates a cgo_library rule for C/C++ code.
+func (g *generator) generateCgoCLib(rel, name string, pkg *build.Package) (*bzl.Rule, error) {
+	kind := "cgo_library"
+
 	attrs := []keyvalue{
 		{key: "name", value: name},
-		{key: "srcs", value: append(pkg.GoFiles, pkg.SFiles...)},
-		{key: "visibility", value: []string{visibility}},
 	}
 
+	srcs := append([]string{}, pkg.CgoFiles...)
+	srcs = append(srcs, pkg.CFiles...)
+	srcs = append(srcs, pkg.CXXFiles...)
+	srcs = append(srcs, pkg.HFiles...)
+	srcs = append(srcs, pkg.SFiles...)
+	attrs = append(attrs, keyvalue{key: "srcs", value: srcs})
+
+	copts := append([]string{}, pkg.CgoCFLAGS...)
+	copts = append(copts, pkg.CgoCPPFLAGS...)
+	copts = append(copts, pkg.CgoCXXFLAGS...)
+	if len(copts) > 0 {
+		attrs = append(attrs, keyvalue{key: "copts", value: copts})
+	}
+	if len(pkg.CgoLDFLAGS) > 0 {
+		attrs = append(attrs, keyvalue{key: "linkopts", value: pkg.CgoLDFLAGS})
```

Comment:
You mean a test different than the ones in generator_test.go?

Is there a high-level "build this repo" test I can steal ideas from in here?